crates/bevy_text/src/pipeline.rs

```diff
@@ -1,219 +1,406 @@
-    }
+fn load_font_to_fontdb(
+    section: &TextSection,
+    font_system: &mut cosmic_text::FontSystem,
+    map_handle_to_font_id: &mut HashMap<AssetId<Font>, (cosmic_text::fontdb::ID, String)>,
+    fonts: &Assets<Font>,
+) {
+    let font_handle = section.style.font.clone();
+    map_handle_to_font_id
+        .entry(font_handle.id())
+        .or_insert_with(|| {
+            let font = fonts.get(font_handle.id()).expect(
+                "Tried getting a font that was not available, probably due to not being loaded yet",
+            );
+            let data = Arc::clone(&font.data);
+            let ids = font_system
+                .db_mut()
+                .load_font_source(cosmic_text::fontdb::Source::Binary(data));
+
+            // TODO: it is assumed this is the right font face
+            let face_id = *ids.last().unwrap();
+            let face = font_system.db().face(face_id).unwrap();
+            let family_name = face.families[0].0.to_owned();
+
+            (face_id, family_name)
+
+            // TODO: below may be required if we need to offset by the baseline (TBC)
+            // see https://github.com/pop-os/cosmic-text/issues/123
+            // let font = font_system.get_font(face_id).unwrap();
+            // map_font_id_to_metrics
+            //     .entry(face_id)
+            //     .or_insert_with(|| font.as_swash().metrics(&[]));
+        });
 }
-impl ToSectionText for TextMeasureSection {
-    #[inline(always)]
-    fn to_section_text(&self) -> SectionText<'_> {
-        SectionText {
-            text: &self.text,
-            scale: PxScale::from(self.scale),
-            font_id: self.font_id,
-        }
-    }
+
+/// Translates [`TextSection`] to [`Attrs`](cosmic_text::attrs::Attrs),
+/// loading fonts into the [`Database`](cosmic_text::fontdb::Database) if required.
+fn get_attrs<'a>(
+    section: &TextSection,
+    section_index: usize,
+    font_system: &mut cosmic_text::FontSystem,
+    map_handle_to_font_id: &'a HashMap<AssetId<Font>, (cosmic_text::fontdb::ID, String)>,
+    scale_factor: f64,
+) -> Attrs<'a> {
+    let (face_id, family_name) = map_handle_to_font_id
+        .get(&section.style.font.id())
+        .expect("Already loaded with load_font_to_fontdb");
+    let face = font_system.db().face(*face_id).unwrap();
+
+    // TODO: validate this is the correct string to extract
+    let attrs = Attrs::new()
+        .metadata(section_index)
+        .family(Family::Name(family_name))
+        .stretch(face.stretch)
+        .style(face.style)
+        .weight(face.weight)
+        .metrics(Metrics::relative(section.style.font_size, 1.2).scale(scale_factor as f32))
+        .color(cosmic_text::Color(section.style.color.to_linear().as_u32()));
+    attrs
+}
+
+/// Calculate the size of the text area for the given buffer.
+fn buffer_dimensions(buffer: &Buffer) -> Vec2 {
+    // TODO: see https://github.com/pop-os/cosmic-text/issues/42 Request: Allow buffer dimensions to be undefined
```

Comment:
The referenced issue was resolved here: https://github.com/pop-os/cosmic-text/pull/271

Reply:
yes, I saw that, but I dont know what that means for this PR? should the TODO just be removed?

Reply:
I am not actually sure how that comment relates to the adjacent code.

@tigregalis , any hints? Sorry if that's not your comment, it's hard to tell from the commit history.

Reply:
from tigregalis on discord:
> We can remove this comment; we do use undefined buffer dimensions when constructing the buffer

Reply:
Resolved by this PR: <https://discord.com/channels/691052431525675048/1248074018612051978/1254285513498562641>

Reply:
We should really be using an undefined buffer size when computing the "max-content" width.